src/groupeddataframe/grouping.jl

```diff
@@ -177,6 +213,318 @@ end
+    print(io, "GroupKey: ")
+    show(io, NamedTuple(k))
+end
+
+Base.parent(key::GroupKey) = getfield(key, :parent)
+Base.length(key::GroupKey) = length(parent(key).cols)
+Base.keys(key::GroupKey) = Tuple(groupvars(parent(key)))
+Base.names(key::GroupKey) = groupvars(parent(key))
+# Private fields are never exposed since they can conflict with column names
+Base.propertynames(key::GroupKey, private::Bool=false) = keys(key)
+Base.values(key::GroupKey) = groupvalues(Tuple, parent(key), getfield(key, :idx))
+
+Base.iterate(key::GroupKey, i::Integer=1) = i <= length(key) ? (key[i], i + 1) : nothing
+
+Base.getindex(key::GroupKey, i::Integer) = groupvalues(parent(key), getfield(key, :idx), i)
+
+function Base.getindex(key::GroupKey, n::Symbol)
+    try
+        return groupvalues(parent(key), getfield(key, :idx), n)
+    catch e
+        throw(KeyError(n))
+    end
+end
+
+function Base.getproperty(key::GroupKey, p::Symbol)
+    try
+        return key[p]
+    catch e
+        throw(ArgumentError("$(typeof(key)) has no property $p"))
+    end
+end
+
+Base.NamedTuple(key::GroupKey) = groupvalues(NamedTuple, parent(key), getfield(key, :idx))
+Base.Tuple(key::GroupKey) = values(key)
+
+
+"""
+    keys(gd::GroupedDataFrame)
+
+Get the set of keys for each group in `GroupedDataFrame` object `gd`, in the
+same order in which the groups appear.
+The keys are [`GroupKey`](@ref) objects which behave like `NamedTuple`s holding
+the values of the grouping keys and can be used to index into `gd` efficiently.
+
+### Returns
+
+Iterator over [`GroupKey`](@ref)s.
+
+### Examples
+
+```jldoctest
+julia> df = DataFrame(a = repeat([:foo, :bar, :baz], outer=[4]),
+                      b = repeat([2, 1], outer=[6]),
+                      c = 1:12);
+
+julia> gd = groupby(df, [:a, :b])
+GroupedDataFrame with 6 groups based on keys: a, b
+First Group (2 rows): a = :foo, b = 2
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+⋮
+Last Group (2 rows): a = :baz, b = 1
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 1     │ 6     │
+│ 2   │ baz    │ 1     │ 12    │
+
+julia> NamedTuple.(keys(gd))
+6-element Array{NamedTuple{(:a, :b),Tuple{Symbol,Int64}},1}:
+ (a = :foo, b = 2)
+ (a = :bar, b = 1)
+ (a = :baz, b = 2)
+ (a = :foo, b = 1)
+ (a = :bar, b = 2)
+ (a = :baz, b = 1)
+
+julia> k = first(keys(gd))
+GroupKey: (a = :foo, b = 2)
+
+julia> keys(k)
+(:a, :b)
+
+julia> values(k)
+(:foo, 2)
+
+julia> NamedTuple(k)
+(a = :foo, b = 2)
+
+julia> k.a
+:foo
+
+julia> k[:a]
+:foo
+
+julia> k[1]
+:foo
+
+julia> gd[k]
+2×3 SubDataFrame
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+```
+"""
+Base.keys(gd::GroupedDataFrame) = (GroupKey(gd, i) for i in 1:length(gd))
```

Comment:
Here (as opposed to `groupvalues`) I think it is OK to be lazy (I am leaving this comment as maybe because we have a lazy option here we can choose to have an eager implementation of `groupvalues`).